[Dong] 자바체스 2일차, step2 Board클래스 개발(재시도)

moto6 opened the pull request:
<내용추가> 
https://github.com/codesquad-members-2021/java-chess/pull/32
- 기존에 step2 에서 잘못 선택한 브랜치로 인해 해당 PR은 Closed 되었었습니다.
- step3와 commit이 섞이는 문제가 있어서, 해당 문제를 해결 후 다시 pr을 올립니다!

@ksundong git pr관련해서 알지 못했던 문제의식을 캐치해주셔서 감사드립니다~

=======================================================


자바체스 2일차, step2 Board클래스 개발에 대한 PR입니다. 

#### 이번 PR에서 변경점은 아래와 같습니다.

- BoardTest.java 파일 추가
    - BoardTEST 파일을 추가하고 루카스 샘플소스코드 추가
    - BoardTEST 파일에서 문제에서 제시된 테스트코드와 코딩컨벤션을 통일함

- Board.java  파일추가
    - Board클래스에서 Pawn 의 추가삭제 함수 시그니쳐만 작성

- Pawn 클래스에
    - 색상을 문자열 처리하던것을 Enum을 사용
    - 색상정보는 생성 이후 상수로 사용 하도록 변경함
    - Pawn의 Color 정보는 상수로 사용

- 프로젝트 관련
    -  파일들을 chess패키지로 이동

이상입니다.

moto6 committed:
Board클래스 문제에 주어진 조건대로 함수 시그니쳐 추가

moto6 committed:
1.함수 시그니쳐 추가
  - char represent 입력을 받는 생성자 추가

2. 입력파라미터가 없는 생성자 변경
 - 이전버전에 비해 인덴트를 지정하고 주석을 추가

3. final color 를 private 키워드를 추가해 접근제한자를 추가함

moto6 committed:
Board 초기화 관련 테스트코드 추가

moto6 committed:
Pawnarr 로 구현하던 기존의 방식을 두개로 나눠서

whitePawn과 balckPawn 으로 나눠서 구현함. 탐색속도 향상을 위해서

moto6 committed:
psvm을 가진 main 클래스 개발 시작함.

auth메서드와 laucher 메서드 개발

moto6 committed:
getWhitePawnsResult,getblackPawnsResult 두가지 메서드를 수정함
-------------------------------------
두 메서드는 90%이상 코드가 중복되는 부분이 있었습니다. 중복되는 부분을 getPawnsResult라는 메서드로 통일하고 기존의 메서드는 새로운 메서드를 래핑하는 방식으로 변경됩니다.

이런 방법을 통해, getPawnsResult 부분의 버그는 해당 메서드만의 문제로 집중할 수 있게 됩니다

moto6 committed:
getSymbol 메서드를 추가합니다.
===================================
이 메서드는 기존에 없었지만, Board 클래스를 수정하면서 필요성이 생겨 추가하였습니다.

이 메서드는 Pawn 인스턴스의 Color 가 화이트인경우 p를 반환하고, 반대의 경우 P를 반환합니다

moto6 committed:
메서드명 단순 수정

moto6 committed:
사용하지않고, 필요하지 않은 주석 삭제

moto6 committed:
원래의 로직클래스 메서드명 변경으로 인한 테스트코드 수정

moto6 committed:
import static 키워드 오류 수정으로 인한 문제 해결
====================================
import static 키워드를 사용해 원래는

Color.WHITE, Color.BLACK 라고 명해야됬던 것들을 WHITE,BLACK 이라고 길이를 줄여서 사용해왔습니다.

이 기법은 객체지향적이지 않고, 코드의 재사용성을 떨어트리는 문제가 있어서 수정합니다

더욱 자세한 사항은 https://unabated.tistory.com/entry/%EC%99%9C-%EC%9E%90%EB%B0%94%EC%97%90%EC%84%9C-static%EC%9D%98-%EC%82%AC%EC%9A%A9%EC%9D%84-%EC%A7%80%EC%96%91%ED%95%B4%EC%95%BC-%ED%95%98%EB%8A%94%EA%B0%80 글을 참고하여 변경하였습니다.

moto6 commented:
@ksundong  리뷰어 지정을 pr 만드는 단계에서 했어야 하는데 미쳐 하지 못했습니다.

늦게나마 맨션으로 대체해도 될까요???

ksundong commented:
@d-h-k 
잠시만 기다려주세용~

ksundong commented on src/main/java/net/Dong/Main.java:
```diff
@@ -1,11 +1,41 @@
 
+
+public class Main {
+    private static final String PROMPT = "J-chess>>";
```
오 프롬프트 ㅎㅎ

moto6 replied:
감사합니다!  아마 코드스쿼드 입학테스트때 해당 변수명으로 만드는게 조건이여서 오마주한 상수명 같습니다 ㅎㅎ..!

ksundong commented on src/main/java/net/Dong/Main.java:
```diff
@@ -1,11 +1,41 @@
+        return;
     }
 
+    public String auth(Scanner sc) {
+        // 멀티플레이, 싱글플레이 구분
+        // 멀티플레이시 플레이어 구분을 위해 필요해서 추가함 추후개발
+        return "";
+    }
```
이 부분은 제거해주세요~

moto6 replied:
넵 실제로 아무일도 하지 않는 메서드는 삭제하겠습니다.

ksundong commented on src/main/java/net/Dong/Main.java:
```diff
@@ -1,11 +1,41 @@
+
 
+
+public class Main {
```
생각해보니 미션2 요구사항에는 Main을 딱히 만들지 않아도 되었던 것 같아요~

moto6 replied:
찬찬히 생각해보니 Main이라는 클래스 자체가 이번 단계에서는 아무런 일을 하지 않습니다. 

미래지향적인 코드는 짜지 말라는 말을 실천에 옴겨 해당 클래스는 삭제하겠습니다!

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -1,31 +1,109 @@
 
 import net.Dong.chess.Pawn;
+import static net.Dong.chess.Pawn.Color;
+//import static net.Dong.chess.Pawn;
```
안쓰는 주석은 제거해주세요~

moto6 replied:
넵 앞으로도 안쓰는 주석문/헤더 특히 import static은 삭제하겠습니다!

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -1,31 +1,109 @@
 
 public class Board {
-    ArrayList<Pawn> arrPawn = new ArrayList<>();
+    ArrayList<Pawn> whitePawn = new ArrayList<>();
+    ArrayList<Pawn> blackPawn = new ArrayList<>();
```
- **접근 제어자** 에 대해서 학습해보시고, 왜 이 코드가 문제가 될 수 있는지 생각해봐주세요!
- 보통 자바에서는 `List<Pawn> whitePawn = new ArrayList<>();` 와 같은 형태를 많이 사용하는데요. 현재단계에서는 문제가 되지 않지만, 다형성이라는 키워드에 대해서 공부해보셨으면 좋겠어요~

ksundong replied:
그리고 보통 colleciton을 네이밍 할 때에는 복수형 표현이나, List라는 접미사를 붙여주는게 좋습니다!

moto6 replied:
말씀해주신 키워드로 찾아보면서 
1.왜 Private 키워드를 사용해야 하는지
2. 왜 ArrayList 를 List형 참조변수에 담아야 하는지
두가지 사항에 대해서 조금이나마 알 수 있었습니다! 감사합니다

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -1,31 +1,109 @@
 
 
-    public int add(Pawn pawn) {
+    public int whiteAdd(Pawn pawn) {
```
저는 addWhite라는 네이밍을 더 선호합니다. 우리가 영작을 할 때 사용하는 순서를 생각하시면 좋아요.

moto6 replied:
네 앞으로도 영작순서를 고려해서 식별자 이름을 정하겠습니다!

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -1,31 +1,109 @@
+        return blackPawn.size();
     }
 
     public void removePawn(Pawn remove) {
         //arrlist에서 삭제하는 코드 작성
     }
```
요구사항에만 맞춰서 개발해봅시다~

moto6 replied:
넵 그러고보니 remove 기능은 요구사항에서 벗어난 듯 합니다. 좋은 말씀 감사드립니다

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -1,31 +1,109 @@
     public int size() {
-        // return number of Mal
-        return arrPawn.size();
+        return whitePawn.size()+blackPawn.size();
```
아래 만들어준 메서드를 활용해도 좋겠네요.

ksundong replied:
```suggestion
        return whitePawnSize() + blackPawnSize();
```

ksundong replied:
변경 요청 드렸는데 반영이 안되어있네요.

moto6 replied:
혹시 
```
public int size() {
        return whitePawn.size() + blackPawn.size();
}
```
위 코드처럼 메서드를 수정하는걸 의도하신건지 궁금합니다 ㅜ.ㅜ....

ksundong replied:
`return whitePawnSize() + blackPawnSize();`
이 형태를 의미했습니다.
이미 아래에 만들어준 메서드를 최대한 활용해봅시다.

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -1,31 +1,109 @@
+        return blackPawn.size();
+    }
+
+    public Pawn findWhitePawn(int index) {
+        return whitePawn.get(index);
+    }
+
+    public Pawn findBlackPawn(int index) {
+        return whitePawn.get(index);
+    }
```
```suggestion
    public Pawn findWhitePawn(int index) {
        return whitePawn.get(index);
    }

    public Pawn findBlackPawn(int index) {
        return blackPawn.get(index);
    }
```

이런 부분은 조심해줘야겠죠?

moto6 replied:
넵 이런 부분은 조심하겠습니다.

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -1,31 +1,109 @@
+    }
+
+    public void initialize() {
+   // white pawn 8, black pawn 8
+        addPawn(Color.BALCK , 8);
+        addPawn(Color.WHITE , 8);
+    }
```
자 8이라는 숫자는 매직넘버입니다. 이를 추출해봅시다.

ksundong replied:
그럼 굳이 주석을 추가해주지 않아도 되겠죠?

moto6 replied:
넵 해당부분을  Pawn클래스가 기본으로 가지고 있는 public static final 값으로 수정하였습니다.

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -1,31 +1,109 @@
+        addPawn(Color.WHITE , 8);
+    }
+
+    private void addPawn(Pawn.Color color, int numOfPawn) {
+        if(color == Color.BALCK) {
+            addPawnBlack(numOfPawn);
+            return;
+        }
+
+        if (color == Color.WHITE) {
+            addPawnBlack(numOfPawn);
+        }
+    }
+
+    private void addPawnWhite(int numOfPawn) {
+        for(int i=0 ; i<numOfPawn ; i++) {
+            whitePawn.add(new Pawn(Color.WHITE));
+        }
+    }
+
+    private void addPawnBlack(int numOfPawn) {
+        for(int i=0 ; i<numOfPawn ; i++) {
+            whitePawn.add(new Pawn(Color.WHITE));
+        }
+    }
```
너무 메서드 추출이 많다고 생각됩니다.

moto6 replied:
넵 말씀해주신 요소를 참고해서 아래처럼 수정했습니다.
- addPawn호출 > addPawnWhite/addPawnBlack :: 기존 2단계에서 
- addPawnWhite/addPawnBlack 호출하는 1단계로 변경하였습니다.

```

    private void addPawnWhite(int numOfPawn) {
        for (int i = 0; i < numOfPawn; i++) {
            whitePawn.add(new Pawn(Color.WHITE));
        }
    }

    private void addPawnBlack(int numOfPawn) {
        for (int i = 0; i < numOfPawn; i++) {
            whitePawn.add(new Pawn(Color.BLACK));
        }
    }

```

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -1,31 +1,109 @@
+
+
+
+    private String print() {
+        //print current state of chess board
+        // 화면에 출력할 결과를 StringBuilder에 저장한 후 String으로 반환하는 메소드를 구현한다
+        StringBuilder sb = new StringBuilder();
+
+        return "";
     }
```
이 부분도 제거해주세요.

moto6 replied:
넵 해당부분 제거하겠습니다.

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -1,31 +1,109 @@
-        // Receives an index and returns it as an object color
-        return arrPawn.get(index).getColor();
+
+    public String getWhitePawnsResult() {
+        return getPawnsResult(whitePawn);
+    }
+
+
+    public String getBlackPawnsResult() {
+        return getPawnsResult(blackPawn);
     }
 
+    private String getPawnsResult(ArrayList<Pawn> whatPawn) {
+        StringBuilder sb = new StringBuilder();
+        int len = whatPawn.size();
+        for(Pawn p : whatPawn) {
+            sb.append(p.getSymbol());
+        }
+        return sb.toString();
+    }
```
3단계 코드가 있는 것 같네요 👀

moto6 replied:
음.. 삭제하면 될까요?? 

- 사실 step3의 커밋들이 step2 에 섞여있는 문제가 있었습니다. 
- 그래서 hard reset과 push --force 명령어로 어느정도 복구는 진행했는데, 미쳐 제거하지 못한 코드들이 있었습니다. 


죄송합니다!

ksundong replied:
너무 지우기 힘드시면 다음단계로 그냥 넘기셔도 됩니다.

ksundong commented on src/main/java/net/Dong/chess/Pawn.java:
```diff
@@ -2,37 +2,54 @@
-    }
 
     public Pawn() {
+        // defalut color : WHITE
```
불필요한 주석은 제거해주세요~

moto6 replied:
넵 앞으로도 불필요한 주석은 제거하겠습니다

ksundong commented on src/main/java/net/Dong/chess/Pawn.java:
```diff
@@ -2,37 +2,54 @@
 
     public enum Color {
         WHITE,
         BALCK
```
```suggestion
        BLACK
```

moto6 replied:
오타가 있었습니다. 이런 실수는 다시 반복되지 않도록 주의하겠습니다!

ksundong commented on src/main/java/net/Dong/chess/Pawn.java:
```diff
@@ -2,37 +2,54 @@
+    }
+
+
     public Pawn(String colorString) {
+        // insert String-type color constructor
         if(colorString.equals("BLACK")) {
             this.color = Color.BALCK;
         }else {
             this.color = Color.WHITE;
         }
     }
```
이 부분은 제거해주셔도 되지 않을까 싶어요~

moto6 replied:
넵 역시나 불필요한 주석이였습니다. 말씀해 주셔서 감사합니다!

ksundong commented on src/main/java/net/Dong/chess/Pawn.java:
```diff
@@ -2,37 +2,54 @@
-    //}
 
-    public Color getColor() {
+    public Pawn(Color color, char represent) {
+        this.color = color;
+    }
```
이 부분도 필요없으니 제거해줍시다!

moto6 replied:
네 해당 생성자는 사용하지 않으니 제거하겠습니다. 

돌이켜보니 생각보다 많은 부분이 불필요한 코드 같습니다. 이부분은 앞으로도 계속 주의하겠습니다.

ksundong commented on src/test/java/net/Dong/BoardTest.java:
```diff
@@ -16,13 +16,22 @@ public void create() throws Exception {
+        Board board = new Board();
+        board.initialize();
+        assertEquals("pppppppp", board.getWhitePawnsResult());
+        assertEquals("PPPPPPPP", board.getBlackPawnsResult());
     }
 }
```
https://github.com/codesquad-members-2021/java-chess/pull/9#discussion_r572027344
이 코멘트를 보시고 고쳐주세요!

moto6 replied:
그동안 잘 알지 못하고 사용했던 부분이네요! 말씀해 주셔서 감사합니다. 
앞으로는 꼭 파일의 끝은 개행으로 끝내고 Tool 차원에서도 실수를 방지하기 위해서

인텔리제이여 Ensure every saved file ends with a line brake 옵션을 활성화 했습니다!

ksundong commented on src/test/java/net/Dong/PawnTest.java:
```diff
@@ -4,17 +4,17 @@
 
 
     @Test
     @DisplayName("TSET >> START ")
```
ㅎㅎ 이런 이름보다는 좀 더 적절한 네이밍을 고민해봅시다.

moto6 replied:
네    DisplayName 에 좀더 적절하게 테스트의 이름을 부여하겠습니다!

ksundong commented on src/test/java/net/Dong/PawnTest.java:
```diff
@@ -4,17 +4,17 @@
     public void create() {
-        verifyPawn(WHITE);
-        verifyPawn(BALCK);
+        verifyPawn(Color.WHITE);
+        verifyPawn(Color.BALCK);
```
`verifyPawn()` 메서드는 assertJ를 사용해서 assertion해주세요!

moto6 replied:
앞으로 모든 testcode에서는 assertJ 관련 api들을 사용하겠습니다!

ksundong commented on src/test/java/net/Dong/BoardTest.java:
```diff
@@ -16,13 +16,22 @@ public void create() throws Exception {
         Board board = new Board();
```
테스트 메서드는 public을 떼줍시다!

moto6 replied:
넵 이번 코드리뷰에서는 접근제어자에 대해서 다시한번 생각해볼 필요성을 느끼고 있습니다. 좋은 말씀 감사드립니다!

ksundong commented on src/test/java/net/Dong/BoardTest.java:
```diff
@@ -16,13 +16,22 @@ public void create() throws Exception {
         assertEquals(1, board.size());
-        assertEquals(Color.WHITE, board.findPawn(0));
+        assertEquals(Color.WHITE, board.findWhitePawn(0));
 
         Pawn black = new Pawn(Color.BALCK);
-        board.add(black);
+        board.blackAdd(black);
         assertEquals(2, board.size());
-        assertEquals(Color.BALCK, board.findPawn(1));
+        assertEquals(Color.BALCK, board.findBlackPawn(1));
```
`assertEquals()` 대신 `assertThat()`을 사용해서 assertion해주세요!

moto6 replied:
넵 ! 감사합니다!

ksundong reviewed (changes requested):
안녕하세요! Dong 리뷰어 디온입니다! step2를 살리느라 고생많으셨어요.
DM 드린대로 Mission3 관련 내용은 따로 제거하시거나 수정하실 필요는 없습니다.
다만, 몇 가지 수정사항이 있어서 변경요청드려요!

### 리뷰 내용

- 코드 포매팅에 신경써주세요.
- 접근 제어자에 대해서 학습하시고, 접근제어자를 알맞게 설정해주세요!
- 주석은 최대한 배제하고 진행해봅시다. 코드만으로 충분히 이해할 수 있도록 작성하는 도전을 꼭 해보세요.
- 불필요하게 메서드 추출을 너무 많이하지 않고, 간단한 로직으로 만드는 것을 먼저 시도해봅시다.
- 당장 사용하지 않는 코드는 모두 제거해봅시다.
- JUnit5에서는 테스트 클래스와 메서드에 `public` 접근제어자는 붙이지 않는 것이 일반적입니다.
- assertJ assertion을 학습하고 이를 이용해봅시다.
- 테스트 명도 테스트가 무엇을 수행하는지 적어주는 것이 좋습니다!

고생많으셨습니다! 수정하시고 다시 Review 요청해주세요!

ksundong commented:
ㅋㅋ 그러고보니 100번째 이슈군요.

ksundong commented:
@d-h-k 
혹시 계속 코드 작성중이신가요? 리뷰 반영해서 푸시해주세요~

moto6 commented:
@ksundong 넵 코드 작성중입니다. 제가 요 몇일 과제에 집중을 못해서 수정사항 반영을 다 하지 못했습니다 죄송합니다..! 오늘(토요일) 저녁 시간 전까지 무조건 리뷰사항 다 반영해서 푸쉬하겠습니다! 

코드리뷰 감사드리고, 푸쉬가 늦어져서 다시한번 늦어져서 죄송합니다

moto6 committed:
사용하지 않는 주석 삭제

moto6 committed:
불필요한 파일이며, 해당 단계에서는 사용하지 않음으로 삭제

moto6 committed:
텍스트파일은 반드시 개행으로 끝을 낸다!

이것은 POSIX표준이므로 지켜주는것이 컨벤션에 적합!

moto6 committed:
해당 코드에는 두가지 문제가 있었습니다.

1. 접근제어자를 사용하지 않아서 ArrayList 참조변수를 외부에서 접근할수 있었습니다.
 - Board 클래스 외부에서 whitePawn 혹은 blackPawn 변수(데이터)에 접근하는 코드가 여러군데서 발생하는 경우, 유지보수, 디버깅 측면에서 불리합니다.
 - 해당 부분을 디폴트에서 private 접근 제어자를 사용하므로써, 해당 ArrayList 데이터에 대한 책임은 Board에서 완전하게 처리해야 합니다.

 2. ArrayList를 담는 List 형 변수
추후 확장을 위해서 ArrayList가 아닌 다른 형태의 List로 변경하더라도 대응할수 있기때문에 확장성이 뛰어나 변경합니다.
- 뿐만 아니라 ArrayList는 List를 상속받은 클래스이므로, 조상클래스로의 Up-Casting은 Implicit casting 을 지원하므로 해당 코드로 변경합니다.

moto6 committed:
assertJ를 사용해서 assertion 진행
이에 따른 import 부분 수정

moto6 committed:
테스트 메서드는 public을 제거해줬습니다.

생각해보니 해당 메서드는 프레임워크에서 호출하고 마는 메서드이므로 public일 필요는 전혀 없습니다.

moto6 committed:
사용하지 않는 import 제거

moto6 committed:
과도하게 복잡한 private 매서드 호출구조 개선

기존 2단계 호출에서 1단계 호출로 개선

moto6 committed:
로직에러 -> 잘못된 상수값 입력 수정

moto6 committed:
테스트 메서드 수정

moto6 committed:
push

moto6 commented:
@ksundong 혹시 새로운 PR을 날려서 merge 받아야 하는 방식이 맞는지 궁금합니다!!

ksundong commented:
@d-h-k 같은 PR에서 리뷰받고 수정하는게 맞는 방식입니다~
올려두신 PR은 main에 PR 요청을 보내기도 하셨고 해서 Close했습니다.

ksundong commented:
커밋메시지에 신경 많이 써주세요~ 나중에 작업 이력이 되는데, `push` 같은 내용이 들어가 있으면 좀 그렇겠죠?

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -1,31 +1,96 @@
 
-    public int add(Pawn pawn) {
+    public int whiteAdd(Pawn pawn) {
+        //arrPawn.add()
```
주석은 제거해주세요~

moto6 replied:
넵 앞으로도 불필요한 주석은 사용하지 않겠습니다.

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -1,31 +1,96 @@
+    public int whiteAdd(Pawn pawn) {
+        //arrPawn.add()
+        whitePawn.add(pawn);
+        return whitePawn.size();
```
이 메서드가 `size()`를 반환하는 이유라도 있나요?

moto6 replied:
size()를 반환하는 이유가 없었습니다.
메서드가 add의 기능만을 충실하게 수행할 수 있도록 수정하겠습니다!

ksundong replied:
추가된 이후의 `size()`를 이용해서 뭔가 처리를 해줄 수도 있어보이긴 했습니다~

ksundong commented on src/test/java/net/Dong/PawnTest.java:
```diff
@@ -1,26 +1,27 @@
     }
 
     void verifyPawn(Color c) {
         Pawn p = new Pawn(c);
-        assertEquals(c, p.getColor(),"fail");
+        assertThat(p.getColor()).isEqualTo(c);
     }
```
p, c 이런 네이밍보다는 pawn, color를 더 자주 사용해주세요.

moto6 replied:
넵 말씀하신 부분을 생각해보니 getcolor보다는 color라는 메서드명이 좀더 의미전달과 가독성에 좋을듯 합니다 

제가 이런 부분에 앞으로 조금 더 주의를 기울이겠습니다

ksundong replied:
음 그걸 의미한건 아닌데요~ `getColor`는 그대로 쓰셔도 될 것 같은데요~

ksundong commented on src/test/java/net/Dong/PawnTest.java:
```diff
@@ -1,26 +1,27 @@
 
 import net.Dong.chess.Pawn;
-import org.junit.jupiter.api.*;
+import net.Dong.chess.Pawn.*;
```
이렇게 import 해주신 이유가 있나요?

moto6 replied:
원래 의도는 Color 클래스를 import 하기위해 *기호로 붙여줬는데요.


기존의 import net.Dong.chess.Pawn.*; 가 아닌


  import net.Dong.chess.Pawn.Color 방식으로 Color class를 import 하겠습니다.
리뷰어님께서 언급해주셔서 생각을 잠깐 해보니 추후 Pawn클래스에 내용이 많아지게 될 경우 컴퓨터에 부담이 갈 것으로 예상됩니다.

제 생각이 미쳐 닿지 못한 부분인데. 잘 말씀해주셔서 감사드립니다 :D

ksundong replied:
컴퓨터에 부담이 갈 정도는 아닙니다.
취향차이긴 한데요 저는 명시적으로 보여주는 것을 선호하는 편이에요~

ksundong commented on src/test/java/net/Dong/BoardTest.java:
```diff
@@ -12,17 +12,26 @@
+
+
+    @Test
+    public void initialize() throws Exception {
+        Board board = new Board();
+        board.initialize();
+        assertThat(board.getWhitePawnsResult()).isEqualTo("pppppppp");
+        assertThat(board.getBlackPawnsResult()).isEqualTo("PPPPPPPP");
     }
```
여기도 public은 떼주세요.
리뷰 받은 부분만 고치는 것이 아니라, 전체 부분을 다 확인해보세요~

moto6 replied:
넵 제가 놓친부분이 있었나 봅니다. 좀더 꼼꼼하게 살펴보겠습니다!

ksundong commented on src/main/java/net/Dong/chess/Pawn.java:
```diff
@@ -2,37 +2,54 @@
+
 
+    public Color getColor() {
+        // return current Pawn's color
```
여기도 주석 제거해줄까요?

moto6 replied:
넵! 바로 위 멘션과 더불어서, 제가 지나치게 성급하고 꼼꼼하지 못한 부분을 짚어주셔서 감사드립니다!

ksundong reviewed (changes requested):
변경 요청 드렸던 부분 다시 확인해서 작업해주세요.
아예 반영이 되어있지 않은 부분도 보이네요.
다음단계로 넘어갈 수 있도록 신경 써주세요!!

moto6 committed:
불필요한 주석 제거

moto6 committed:
add 메서드에 어울리지 않는 동작 수정 -    int 형 size정보를 리턴

moto6 committed:
add 메서드에 어울리지 않는 동작 수정

 int 형 size정보를 리턴하던 기존 동작에서
 void 형으로 변경

 조금더 메서드명과 실제 메서드의 행위가 일치됩니다

moto6 committed:
불필요한 주석 삭제

moto6 committed:
세미콜론 추가

moto6 committed:
메서드명 변경 기존getColor >> color

 의미전달과 가독성에 좋은 메서드의 이름으로 변경

moto6 committed:
Color class import 방식을 변경

기존의 import에서 * 기호를 사용하는 경우 Pawn클래스에 내용이 많아지는 경우 컴퓨터에 부담이 가고, 프로젝트의 스케일업에 따라 사이드 이펙트의 위험성이 있어 수정합니다.

moto6 committed:
불필요한 주석들을 제거합니다

moto6 committed:
테스트 로직에서 public을 제거합니다

moto6 committed:
오타수정 BALCK >> BLACK

ksundong reviewed (changes requested):
수정요청드린 부분이 있는데 수정이 안되어있습니다.
리뷰사항 처음부터 확인좀 해주세요 ㅠㅠ
리뷰사항 확인하시고 수정하시면 mark as resolved를 선택하셔서 처리된 부분을 확인해주세요.

ksundong commented on src/test/java/net/Dong/PawnTest.java:
```diff
@@ -1,26 +1,25 @@
     }
 
     void verifyPawn(Color c) {
         Pawn p = new Pawn(c);
-        assertEquals(c, p.getColor(),"fail");
+        assertThat(p.color()).isEqualTo(c);
     }
```
`p`, `c`를 `pawn`, `color`와 같이 변경해주세요.
`getColor()` 는 그대로 두셔도됩니다.

moto6 replied:
넵 말씀하신 부분 수정완료 하였습니다!

moto6 committed:
변수명에는 축약어를 사용하지 않는다

moto6 committed:
코드 포멧팅 통일

moto6 committed:
코드 포메팅 통일 및 불필요한 개행 제거

moto6 committed:
불필요한 개행 제거

moto6 committed:
enum의 세미콜론 위치 변경

moto6 committed:
Pawn 클래스의 맴버 메서드명 변경
pawn.color 에서 pawn.getColor으로 변경

moto6 committed:
테스트함수의 이름을 수정

테스트의 이름이 테스트가 구체적으로 무엇을 테스트하는지를 더 잘 나타내게 됩니다

moto6 committed:
테스트 명 수정

테스트함수의 이름이, 조금더 정확하게 테스트하는 대상을 의미하도록 변경됩니다.

moto6 committed:
불필요한  throws Exception 제거

moto6 committed:
조금 더 적절한 변수명으로 변경

moto6 committed:
메서드명 변경으로 인한 테스트코드 변경

moto6 committed:
요구사항에 없는 remove 기능은 삭제

moto6 committed:
기능이 없는 메서드는 제거

moto6 committed:
불필요하고 사용하지 않는 생성자 제거

moto6 committed:
과도한 공백 제거

moto6 committed:
8개의 Pawn을 추가하는 부분은 게임의 Rule에 해당하므로, 상수값을 클래스에 내장함

moto6 commented:
안녕하세요 @ksundong 

두가지정도가 제가 판단이 잘 안서서 추가 질문 드리고 싶습니다!

### 1. size 메서드 수정의 건
https://github.com/codesquad-members-2021/java-chess/pull/100#discussion_r579642021
https://github.com/codesquad-members-2021/java-chess/pull/100#discussion_r577538748
- 위 두가지 discussion에서 의도하신바가 size 메서드에서 
```
public int size() {
        return whitePawn.size() + blackPawn.size();
}
```
- 위 코드로의 수정을 의미하신다고 판단했습니다. 혹시 제가 잘못 이해하고 있는지 여쭤보고 싶습니다!



### 2. step3 코드 섞여있는 부분
https://github.com/codesquad-members-2021/java-chess/pull/100#discussion_r577541066
-  step3 코드가 섞여있는 부분이 있는데, 제거를 의도하신게 맞는지 궁금해서 확인 한번만 부탁드리겠습니다!

### 마지막으로
@ksundong  꼼꼼하게 코드리뷰해주셔서 정말정말 진심으로 깊은 감사 드립니다 ㅠ..ㅠ!!

이번에 특히나 github에 "Load more…" 버튼과 "resolve conversation"  기능에 대해서도 처음 알게 되었는데 친절하게 가이드 해주셔서 두번 감사드립니다!

ksundong commented:
@d-h-k 
1. 해당 내용은 코멘트에 다시 작성해두었습니다!
2. step3 코드는 제거하기 쉬운 부분은 바로 걷어내주시고, 걷어내기 어려운 부분은 그냥 남겨두세요!

ㅎㅎ 열심히 해주셔서 저야말로 감사드립니다!

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -1,31 +1,77 @@
+
+    private void addPawnBlack(int numOfPawn) {
+        for (int i = 0; i < numOfPawn; i++) {
+            whitePawn.add(new Pawn(Color.BLACK));
```
여기는 아직도 blackPawn이 아니군요... 😱

ksundong replied:
꼭 테스트 코드 추가하셔서 이 부분 확실하게 보완해주세요!!

moto6 replied:
넵 아래 테스트 코드에서 blackpawn 과 whitepawn 이 혼동되는 부분을 테스트했습니다 

```
    @Test
    void BoardPawnCreate() {
        Board board = new Board();

        Pawn white = new Pawn(Color.WHITE);
        board.addWhitePawn(white);
        assertThat(board.size()).isEqualTo(1);
        assertThat(board.findWhitePawn(0)).isEqualTo(white);

        Pawn black = new Pawn(Color.BLACK);
        board.addBlackPawn(black);
        assertThat(board.size()).isEqualTo(2);
        assertThat(board.findBlackPawn(0)).isEqualTo(black);
    }

```

moto6 replied:
꼼꼼하게 리뷰해주셔서 정말정말 감사합니다!

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -1,31 +1,77 @@
+
+    private String getPawnsResult(List<Pawn> whatPawn) {
+        StringBuilder sb = new StringBuilder();
+        int len = whatPawn.size();
```
필요없는 변수가 아닐까 싶네요~

moto6 replied:
넵 해당부분에 필요없는 변수를 삭제하겠습니다

ksundong reviewed (approved):
고생하셨습니다!
조금은 빡빡하게 리뷰를 했는데, 이 부분을 정말 좋게 봐주셔서 감사해요~
버그가 될 수 있는 부분이 남아있지만, 다음단계를 진행하실 때, 꼭 테스트 코드 작성하시고, 고쳐주신 다음 진행해주세요!

honux77 commented:
수고 많으셨어요!
